pkg/kms/webkms/remotekms.go

```diff
@@ -0,0 +1,297 @@
+
+	return kid, kidURL, nil
+}
+
+// Get key handle for the given KeyID remotely
+// Returns:
+//  - handle instance representing a remote keystore URL including KeyID
+//  - error if failure
+func (r *RemoteKMS) Get(keyID string) (interface{}, error) {
+	return r.buildKIDURL(keyID)
+}
+
+func (r *RemoteKMS) buildKIDURL(keyID string) (string, error) {
+	keystoreURL, err := r.configStore.Get(KeystoreURLField)
+	if err != nil {
+		return "", err
+	}
+
+	return string(keystoreURL) + "/keys/" + keyID, nil
+}
+
+// Rotate remotely a key referenced by KeyID and return a new handle of a keyset including old key and
+// new key with type kt. It also returns the updated KeyID as the first return value
+// Returns:
+//  - new KeyID
+//  - handle instance (to private key)
+//  - error if failure
+func (r *RemoteKMS) Rotate(kt kms.KeyType, keyID string) (string, interface{}, error) {
+	return "", nil, errors.New("function Rotate is not implemented in remoteKMS")
+}
+
+// ExportPubKeyBytes will remotely fetch a key referenced by id then gets its public key in raw bytes and returns it.
+// The key must be an asymmetric key.
+// Returns:
+//  - marshalled public key []byte
+//  - error if it fails to export the public key bytes
+func (r *RemoteKMS) ExportPubKeyBytes(keyID string) ([]byte, error) {
+	keyURL, err := r.buildKIDURL(keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	destination := keyURL + "/keys/export"
+
+	httpReq := &exportKeyReq{
+		KeyID:      keyID,
+		Passphrase: r.secret,
+	}
+
+	req, err := json.Marshal(httpReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal ExportPubKeyBytes key request [%s, %w]", destination, err)
+	}
+
+	resp, err := r.httpClient.Post(destination, ContentType, bytes.NewBuffer(req))
+	if err != nil {
+		return nil, fmt.Errorf("posting ExportPubKeyBytes key failed [%s, %w]", destination, err)
+	}
+
+	// handle response
+	defer closeResponseBody(resp.Body, logger, "ExportPubKeyBytes")
+
+	keyBytes, err := base64.URLEncoding.DecodeString(resp.Header.Get(KeyBytesHeader))
+	if err != nil {
+		return nil, err
+	}
+
+	return keyBytes, nil
+}
+
+// CreateAndExportPubKeyBytes will remotely create a key of type kt and export its public key in raw bytes and returns
+// it. The key must be an asymmetric key.
+// Returns:
+//  - KeyID of the new handle created.
+//  - marshalled public key []byte
+//  - error if it fails to export the public key bytes
+func (r *RemoteKMS) CreateAndExportPubKeyBytes(kt kms.KeyType) (string, []byte, error) {
+	kid, _, err := r.Create(kt)
+	if err != nil {
+		return "", nil, err
+	}
+
+	pubKey, err := r.ExportPubKeyBytes(kid)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return kid, pubKey, nil
+}
+
+// PubKeyBytesToHandle is not implemented in remoteKMS.
+func (r *RemoteKMS) PubKeyBytesToHandle(pubKey []byte, kt kms.KeyType) (interface{}, error) {
+	return nil, errors.New("function PubKeyBytesToHandle is not implemented in remoteKMS")
```

Comment:
`PubKeyBytesToHandle` is being used by the zcap primitives: https://github.com/trustbloc/edge-core/blob/8631b9765a2ef8aea7d2280e2c130ef820d2c78a/pkg/zcapld/middleware.go#L112-L111

The use case is dereferencing a `did:key` URL as a verificationMethod and using `crypto.Crypto` to verify, and it needs a key handle.

The `did:key` URL is not equal to the KMS' internal KID.

Reply:
Implementation of this function should be done in a separate pr, it needs more thinking, ie: how to resolve `did:key` URL to key server URL..

also, do we want to tie remoteKMS to Tink api, I tried to avoid this coupling as much as possible.. to be discussed

Reply:
I assume that we are going to use `ExportPubKeyBytes` and then "manually" build `did:key` from the public key bytes

Reply:
probably `ExportPubKeyBytes` can be implemented as matching pubkey to the one on remote KMS and then returning handle of that key that in case of remote KMS is key URL

Reply:
`ExportPubKeyBytes` should just return raw public key bytes, it should not manage converting its format.

Also if you notice, `Create()` function doesn't return the Tink `*keyset.Handle` in the second return statement, it returns the key URL as defined by the key server instead.

Similarly, IMHO, to create a `*keyset.Handle` should be managed outside of remoteKMS

Reply:
>probably ExportPubKeyBytes can be implemented as matching pubkey to the one on remote KMS and then returning handle of that key that in case of remote KMS is key URL

@aholovko correct, and the crypto service would be a `remoteCrypto` instance that can take the key URL as the key handle argument in `crypto.Verify()`